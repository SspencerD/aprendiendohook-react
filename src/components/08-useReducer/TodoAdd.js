import React from 'react'
import { useForm } from '../../Hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [{ description }, handleinputChange, reset] = useForm({
        description: ''

    });


    const handleSubmit =(e)=>{
        e.preventDefault();

        if (description.trim().length <= 1) {

            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddTodo(newTodo);
        reset();

    }

    return (
        <>

            <h4>Agregar</h4>
            <hr />
            <form className="mb-3" onSubmit={handleSubmit}>
                <input
                    className="form-control form-control-sm"
                    type="text"
                    name="description"
                    placeholder="Aprender algo..."
                    autocomplete="off"
                    value={description}
                    onChange={handleinputChange}
                />
                <button
                    className="btn btn-sm btn-outline-success mt-2 btn-block"
                    type="submit"
                >
                    Agregar
                        </button>
            </form>

        </>
    )
}
