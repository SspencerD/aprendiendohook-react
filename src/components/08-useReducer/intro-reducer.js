
const initialState = [{
    id:1,
    todo: 'Estudiar mucho react',
    done: false
}];

const toDo = (state = initialState, action) => {

    if( action?.type=== 'agregar'){
        // como no se ocupa push , se puede ocupar el operador spread
        // ya que llama a la lista anterior y lo agrega
        return [...state , action.payload ];
    }
    //solo puede tener una sola condici√≥n , no se puede agregar un else
    // y otro return ya que estaria en contra del reducer

    return state;
}

let todos = toDo();

// No usar push si estan usando react (reducer)
// ya que modifica  o muta el obj

const newTodo ={
    id:2,
    todo: 'Estudiar mucho redux',
    done: false

}

const addAction ={

    // payload nombre de como se llama al nuevo objeto

    type: 'agregar',
    payload: newTodo
}

todos = toDo(todos, addAction );

console.log(todos);