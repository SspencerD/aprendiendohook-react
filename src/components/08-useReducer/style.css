body{
    padding: 50px;
}

p{
    cursor: pointer;
    margin: 0 px;
}

h1{
    align-items: center;
}

.list-group-item{
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.badge{
    padding-right:  -10px;
    border-radius: 21px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}
