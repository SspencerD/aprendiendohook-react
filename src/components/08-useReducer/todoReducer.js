

export const todoReducer = (state = [], action) => {

    switch (action.type) {
        case 'add':

            return [...state, action.payload];

        case 'delete':

            return state.filter(todo => todo.id !== action.payload); // seria el grupo id

        case 'toggle': //metodo abreviado

        return state.map( todo =>
            ( todo.id === action.payload)
            ? {...todo, done: !todo.done}
            :todo );
        
            case 'toggle-old': //metodo muy largo de hacerlo

            return state.map(todo => {

                if (todo.id === action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                }else {
                    return todo;
                }

            
            })

        default:
return state;
    }


}