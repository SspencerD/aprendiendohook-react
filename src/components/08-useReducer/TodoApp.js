import React, { useEffect, useReducer } from 'react'
import './style.css';
import { todoReducer } from './todoReducer';
import { Todolist } from './Todolist';
import { TodoAdd } from './TodoAdd';



const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];


    // Esto estamos haciendo al momento de renderizar el init
    // return [{
    //     id: new Date().getTime(),
    // desc: 'Aprender React',
    // done: false
    // }];

}
export const TodoApp = () => {


    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {

        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos]);

    const handleDelete = (todoId) => {

        const borrar = {
            type: 'delete',
            payload: todoId
        };

        dispatch(borrar);

    }

    const handleToggle = (todoId) => {

        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo

        });
    }



    return (
        <div>

            <h1 className="alert alert-dark">To Do APP <small className="badge bg-info"> {todos.length} </small> </h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <Todolist
                        todos={todos}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    />
                </div>

                <div className="col-5">

                    <TodoAdd handleAddTodo={ handleAddTodo } />
                   
                </div>
            </div>
        </div>


    )
}
