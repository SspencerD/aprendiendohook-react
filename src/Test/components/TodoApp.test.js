
import React from 'react';
import {shallow, mount} from 'enzyme';
import { TodoApp } from '../../components/08-useReducer/TodoApp';
import { demoTodo } from '../Data/datatodos';
import { act  } from '@testing-library/react';


describe('Pruebas en <TodoApp/>', () => {
    

    const wrapper = shallow(<TodoApp/>);

    Storage.prototype.setItem = jest.fn(() => {});

    test('Debe de mostrarse correctmente', () => {
        
        expect(wrapper).toMatchSnapshot();
    });


    test('Debe agregar un ToDo', () => {

        // mount viene de enzyme y sirve para 
        //probar toda la app en general

        const wrapper = mount( <TodoApp/>);

        act( () =>{

            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[1]);
        });

        expect( wrapper.find('h1').text().trim()).toBe('To Do APP  2');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('Debe de eliminar un To Do', () => {

        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[0]);
        expect(wrapper.find('h1').text().trim() ).toBe('To Do APP  1');

        wrapper.find('Todolist').prop('handleDelete')(demoTodo[0].id);
        expect(wrapper.find('h1').text().trim() ).toBe('To Do APP  0');


        
    });
});