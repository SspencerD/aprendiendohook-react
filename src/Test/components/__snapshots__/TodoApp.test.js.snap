// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> Debe de mostrarse correctmente 1`] = `
<div>
  <h1
    className="alert alert-dark"
  >
    To Do APP 
    <small
      className="badge bg-info"
    >
       
      0
       
    </small>
     
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <Todolist
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
